### ISSUE TITLE
Reddit.com doesn't load completely within browser.html

### ISSUE THREAD

**Issue by paulrouget** [servo/servo#11223]
Duplicate of https://github.com/browserhtml/browserhtml/issues/1037 as this is apparently a servo bug.

Often the mozbrowser iframe doesn't get the `loadend` event when loading reddit.com.

It appears to happen only with browserhtml.

This is what I managed to figure out so far:

- In document.rs, in `finish_load`, for the `reddit.com` pipeline, `DocumentLoadsComplete` is properly sent. Which means that `loader.is_blocked()` return `false`.
- then, in `script_thread.rs`, in `handle_loads_complete`, `doc.loader().is_blocked()` returns `true`, preventing finishing the load.

So I guess we need to figure out why sometimes the loader gets blocked.

**Comment by paulrouget** [comment]
Just to confirm, the loaders we are talking about in `finish_load` and in `handle_loads_complete` (which comes after), are both the same (same pipeline).

**Comment by paulrouget** [comment]
This script is added:

`https://engine.a.redditmedia.com/ados?t=1463476273520&request={%22Placements%22:[{%22A%22:5146,%22S%22:24950,%22D%22:%22main%22,%22AT%22:5,%22Properties%22:{}},{%22A%22:5146,%22S%22:24950,%22D%22:%22sponsorship%22,%22AT%22:8,%22Properties%22:{}}],%22Keywords%22:%22s.frontpage%2Cs.loggedout%2Cs.sfw%22,%22Referrer%22:%22undefined%22,%22IsAsync%22:true,%22WriteResults%22:true}`

… and never removed from the loader list.

**Comment by paulrouget** [comment]
So we never get the headers for this script. So I guess it's a network issue.

**Comment by paulrouget** [comment]
I'm wondering if I'm hitting the ssl issue of https://github.com/servo/servo/issues/11008

I'm not getting headers (at least, `headers_available` is not called in `htmlscriptelement.rs`) but still, this script is added to the loader. Any idea what else I could look at?

**Comment by tschneidereit** [comment]
This seems to only happen on slower machines, or with fewer CPU cores: on a 2012 MBP with 4 (physical) cores, I can't reproduce at all, whereas on a 2015 Macbook with 2 cores, it happens reliably.

**Comment by metajack** [comment]
@tschneidereit have you tried rr's random scheduler?

@nox This might be a good bug to take.

**Comment by nox** [comment]
Will look into that.

**Comment by paulrouget** [comment]
CPU usage hits ~100% while that happens.

I'll attempt to profile, that might help us understand if we get stuck somewhere.

**Comment by paulrouget** [comment]
Also - depending on time, reddit doesn't serve the same ads, so it might not happen all the time. At the moment, it sill happens, but with a different URL.

**Comment by paulrouget** [comment]
Result of the profile *after* the loadend got blocked: http://people.mozilla.org/~prouget/bugs/issue11223.svg (csv version: http://people.mozilla.org/~prouget/bugs/issue11223.csv).

**Comment by paulrouget** [comment]
@till / @gozala can you guys check (or tell me how to check) why we have JS code running while the iframe is not sending event anymore?

PS: I'm disabling the progressbar like this, is that enough?

```diff
diff --git a/src/browser/web-view.js b/src/browser/web-view.js
index a20291d..5e400d0 100644
--- a/src/browser/web-view.js
+++ b/src/browser/web-view.js
@@ -810,7 +810,7 @@ export const view =
           )
         ]
       )
-    , Progress.view(model.progress, address)
+    // , Progress.view(model.progress, address)
     , html.div
       ( { className: 'webview-tab-icon'
         , style:
```

**Comment by paulrouget** [comment]
I think it's also necessary to remove `ProgressAction(Progress.Start(time))` from `startLoad` to not have an JS running.

**Comment by paulrouget** [comment]
Commenting out `ProgressAction(Progress.Start(time))` (code that updates the progressbar while the page is loading) in browserhtml makes the issue disappear.

Using `setTimeout(16)` instead of `requestAnimationFrame` also makes the issue disappear (we know that `rAF` has an overhead: https://github.com/servo/servo/issues/9844).

So basically, when CPU gets busy, a script might never be executed/downloaded (not sure yet) and blocks the document loader. That also explains why we can not reproduce this issue on a Macbook Pro (faster CPU).

@nox, is there anything I can do to help profiling/debugging this?

**Comment by nox** [comment]
@paulrouget Could we somehow list the blocking loads stored in `DocumentLoader::blocking_loads`?

**Comment by nox** [comment]
Do you think we can reproduce it by using `nice` on Servo?

**Comment by nox** [comment]
Is the HTML sample in https://github.com/browserhtml/browserhtml/issues/1037 still relevant? Seems like it happens 100% of the time when I try locally with it.

**Comment by nox** [comment]
Oh, not 100% of the time, but I can reproduce it on my fancy laptop at least.

**Comment by nox** [comment]
Actually I had just missed the `end` message the first few times, so I didn't notice I wasn't reproducing the never-ending load. :(

**Comment by paulrouget** [comment]
> Could we somehow list the blocking loads stored in DocumentLoader::blocking_loads?

Not sure to understand the question. You want to `println` them?

> Do you think we can reproduce it on a better computer by using nice on Servo?

I haven't tried, but I guess if you nice-up the the process or somehow manage to limit the number of available cores, that might help.

To reproduce, something "heavy" needs to happen on the side. browserhtml happens to have an expensive `rAF` loop. Maybe we can have a reduced test case if we re-use the example from browserhtml/browserhtml#1037 with a random busy script running at the same time.

**Comment by paulrouget** [comment]
> To reproduce, something "heavy" needs to happen on the side. browserhtml happens to have an expensive rAF loop. Maybe we can have a reduced test case if we re-use the example from browserhtml/browserhtml#1037 with a random busy script running at the same time.

Just tried. Very expensive `rAF`. I can't reproduce outside of browserhtml.

**Comment by pcwalton** [comment]
@nox Mind if I assign you here?

**Comment by nox** [comment]
@pcwalton Done.

**Comment by paulrouget** [comment]
So for the script mentioned in my 2nd comment, headers are being downloaded.

I can tell because the script go through this: https://github.com/servo/servo/blob/dea610986dbf4a1e5898cbe6b57832784e786f3a/components/net/resource_thread.rs#L145 (`start_sending_opt`).

But I think the listener is not called. I added a `println` here https://github.com/servo/servo/blob/dea610986dbf4a1e5898cbe6b57832784e786f3a/components/net_traits/lib.rs#L155 (`ResponseAction::process`) and I don't see anything for this specific script.

2 questions:

1. Am I right to think that `ResponseAction::process` is supposed to be called after `start_sending_opt`?
2. If so, that means the mechanism that call the listener doesn't work. I see that we use an `IpcSender` for that. So could that be an IPC issue? How do I debug that?

**Comment by paulrouget** [comment]
Ignore previous comment. I'm getting `response_complete`.

**Comment by paulrouget** [comment]
The script is being downloaded via `http_loader`. In [`ProgressSender::send`](https://github.com/servo/servo/blob/master/components/net/resource_thread.rs#L58)(*resource_thread.rs*) eventually get the complete response. Calls `invoke_with_listener` (*net_traits/lib.rs*), which calls [`sender.send(action)`](https://github.com/servo/servo/blob/dea610986dbf4a1e5898cbe6b57832784e786f3a/components/net_traits/lib.rs#L171). But nothing ever happens after that.

`sender` is an `IpcSender<>`. So looking at [ipc-channel](https://github.com/servo/ipc-channel), once the script is fully downloaded and `sender.send(action)` is called, the method [`IpcSender::send`](https://github.com/servo/ipc-channel/blob/7f10e286a80ea908d1e0d0d115894fd00055e2cb/ipc.rs#L143) is called again and again and again. Never stopping.

I know nothing about IPC, but I guess that's not good.

**Comment by nox** [comment]
@paulrouget Managed to reproduce it on my MBP by limiting the number of cores in Instruments.app's preferences. I never see an 'end' message anymore, though it doesn't use 100% CPU.

**Comment by nox** [comment]
It seems I can't reproduce it since #11274 landed. That included an SSL fix in Hyper.

**Comment by nox** [comment]
And that's probably unrelated, given @paulrouget can still reproduce it. :(

**Comment by bobthekingofegypt** [comment]
I can reproduce it, can't understand it yet though.

To reproduce it on linux (sorry don't use OSX if I can help it) I just make my computer really busy doing pointless stuff.

In one terminal I run a bash script
```
while true; do
    echo "something pointless"
done
```

start it up with

```
taskset -c 1 ./slowmedown.sh
```

Taskset will set the cpu affinity bonding you to only core 1, the script will just spin round using 100% cpu.

Then just start servo with taskset, and whatever parameters you like.
```
taskset -c 1 ./mach run -r -w --pref dom.mozbrowser.enabled a.html 2>&1 | ts > log.txt
```

For me the end event never fires (it's not in the log.txt) when the slowmedown script is running and fires everytime when it isn't.  It would take me a lot longer to actually figure out why that is.  As paulrouget found it always appears to be linked to https://engine.a.redditmedia.com/ados, perhaps because it's a script tag inside an iframe inside an iframe.  For me it all falls down on the events that would free the script to begin downloading:
https://github.com/servo/servo/blob/master/components/net/http_loader.rs#L817
This should be no-op round trip at the moment, as it appears to be ground work for service workers.  It appears the data makes it over ipc but in the script_threads loop the message never gets processed so the None response is never sent back to free the http_loader.  Too much code I've never looked at to fully understand what is happening and why though.

**Comment by paulrouget** [comment]
@bobthekingofegypt thank you, that's useful. Do you reproduce with browserhtml or without?

**Comment by bobthekingofegypt** [comment]
@paulrouget I could reproduce it with and without.

If I have time, and it's still happening, I'm happy to have a look into it next week.

**Comment by bobthekingofegypt** [comment]
I spent a bit of time this afternoon playing around with this issue, but I think I've reached the limit of my current knowledge of Servo code.  I could just be chasing ghosts but it might be useful.

So reddit's page is pretty complicated; it contains many iframes, that contain other iframes, it also has timers that load more scripts and change things in those iframes.
If I run the page on a heavily loaded computer everything appears to fail in a very consistent way.  By fail I just mean not successfully complete the loading of the site.

At the point it fails you have 4 script threads looping:

* thread 0 - a.html
* thread 1 - reddit iframe
* thread 2 - gtm_jail
* thread 3 - ad display iframe

Thread 3 is where the problem comes in, just after the ad-display iframe does a javascript insertion of a new script element (the ados.js file) the script_thread will process a DOMContentLoaded event (as the basic part of the iframe page is loaded).  This event will go all the way to here - https://github.com/servo/servo/blob/master/components/script/dom/eventtarget.rs#L151 but it will never return from that call and I don't know who the listener was.  In the background there are still more messages being delivered to message receivers for script thread 3 but they are never processed because it is stuck.

I don't know where to look next as I don't understand what happens with the codegen stuff or why it would be blocking in there.  It's also pretty hard to debug as there is so much going on all over the place.  But it was interesting to follow the code through the different sections.

**Comment by jdm** [comment]
That's a really good lead! The call to `listener.HandleEvent_` would end up in the `HandleEvent` method of `EventListener` in `EventListenerBinding.rs` under `target/debug/build/script-[hash]/out/Bindings/`. We could step into `JS_CallFunctionValue` from there and potentially figure out the JS code being executed by calling `js::DumpBacktrace(cx)` at some point.

**Comment by jdm** [comment]
Doing so leads me to https://www.redditstatic.com/ad-dependencies.5O7sMdAReBw.js but I don't know if I can trust anything about the line and column information, since Servo routinely fudges those.

**Comment by nox** [comment]
That file seems to be the only interesting one that adds a listener for `DOMContentLoad`.

**Comment by jdm** [comment]
My current theory is that we end up in an endless loop in JS, since we never return from JS::Interpret.

**Comment by jdm** [comment]
Here's a backtrace of the thread that doesn't return from the JS interpreter after killing the process:
```
(lldb) bt
* thread #103: tid = 0x9ef6d, 0x00007fff8e230746 libsystem_kernel.dylib`__psynch_mutexwait + 10
  * frame #0: 0x00007fff8e230746 libsystem_kernel.dylib`__psynch_mutexwait + 10
    frame #1: 0x00007fff84e38779 libsystem_pthread.dylib`_pthread_mutex_lock + 372
    frame #2: 0x0000000100dde852 servo`std::sys::mutex::{{impl}}::lock(self=0x000000011b06bb40) + 34 at mutex.rs:34
    frame #3: 0x0000000100dde825 servo`std::sys_common::mutex::{{impl}}::lock(self=0x000000011b06bb40) + 21 at mutex.rs:35
    frame #4: 0x0000000100fa3126 servo`std::sync::mutex::{{impl}}::lock<layout::layout_thread::LayoutThreadData>(self=0x000000011b091010) + 38 at mutex.rs:224
    frame #5: 0x0000000101040164 servo`layout::query::{{impl}}::offset_parent(self=0x000000011b014290) + 68 at query.rs:139
    frame #6: 0x0000000101eeac35 servo`script::dom::window::{{impl}}::offset_parent_query(self=0x0000000126832800, node=TrustedNodeAddress at 0x000000013b552828) + 117 at window.rs:1226
    frame #7: 0x00000001018e2aed servo`script::dom::htmlelement::{{impl}}::OffsetHeight(self=0x000000012689bc00) + 189 at htmlelement.rs:325
    frame #8: 0x00000001018e29e0 servo`script::dom::bindings::codegen::Bindings::HTMLElementBinding::get_offsetHeight(cx=0x000000010b0d5240, _obj=Handle<*mut js::jsapi::JSObject> at 0x000000013b552980, this=0x000000012689bc00, args=JSJitGetterCallArgs at 0x000000013b552970) + 48 at HTMLElementBinding.rs:546
    frame #9: 0x00000001023412e9 servo`CallJitGetterOp(info=0x0000000104e0ed70, cx=0x000000010b0d5240, thisObj=JS::HandleObject at 0x000000013b552a08, specializedThis=0x000000012689bc00, argc=0, vp=0x000000013b552f68) + 121 at jsglue.cpp:445
    frame #10: 0x000000010168f735 servo`script::dom::bindings::utils::generic_call(cx=0x000000010b0d5240, argc=0, vp=0x000000013b552f68, is_lenient=false, call=0x0000000102341270) + 1397 at utils.rs:482
    frame #11: 0x00000001016908dc servo`script::dom::bindings::utils::generic_getter(cx=0x000000010b0d5240, argc=0, vp=0x000000013b552f68) + 44 at utils.rs:498
    frame #12: 0x000000010337c24b servo`js::CallJSNative(cx=0x000000010b0d5240, native=0x00000001016908b0, args=0x000000013b552f10)(JSContext*, unsigned int, JS::Value*), JS::CallArgs const&) + 187 at jscntxtinlines.h:235
    frame #13: 0x000000010333b913 servo`js::Invoke(cx=0x000000010b0d5240, args=0x000000013b552f10, construct=NO_CONSTRUCT) + 1091 at Interpreter.cpp:476
    frame #14: 0x000000010336682d servo`js::Invoke(cx=0x000000010b0d5240, thisv=0x000000013b553550, fval=0x000000013b553000, argc=0, argv=0x0000000000000000, rval=JS::MutableHandleValue at 0x000000013b552f00) + 621 at Interpreter.cpp:528
    frame #15: 0x0000000103367235 servo`js::InvokeGetter(cx=0x000000010b0d5240, thisv=0x000000013b553550, fval=Value at 0x000000013b553000, rval=JS::MutableHandleValue at 0x000000013b552ff8) + 149 at Interpreter.cpp:640
    frame #16: 0x00000001033adbd6 servo`CallGetter(cx=0x000000010b0d5240, obj=JS::HandleObject at 0x000000013b5530b0, receiver=JS::HandleValue at 0x000000013b5530a8, shape=js::HandleShape at 0x000000013b5530a0, vp=JS::MutableHandleValue at 0x000000013b553098) + 246 at NativeObject.cpp:1735
    frame #17: 0x00000001033a146d servo`bool GetExistingProperty<(cx=0x000000010b0d5240, receiver=js::MaybeRooted<JS::Value, js::AllowGC>::HandleType at 0x000000013b5531b0, obj=js::MaybeRooted<js::NativeObject *, js::AllowGC>::HandleType at 0x000000013b5531a8, shape=js::MaybeRooted<js::Shape *, js::AllowGC>::HandleType at 0x000000013b5531a0, vp=js::MaybeRooted<JS::Value, js::AllowGC>::MutableHandleType at 0x000000013b553198)1>(JSContext*, js::MaybeRooted<JS::Value, (js::AllowGC)1>::HandleType, js::MaybeRooted<js::NativeObject*, (js::AllowGC)1>::HandleType, js::MaybeRooted<js::Shape*, (js::AllowGC)1>::HandleType, js::MaybeRooted<JS::Value, (js::AllowGC)1>::MutableHandleType) + 877 at NativeObject.cpp:1783
    frame #18: 0x00000001033a17ba servo`bool NativeGetPropertyInline<(cx=0x000000010b0d5240, obj=js::MaybeRooted<js::NativeObject *, js::AllowGC>::HandleType at 0x000000013b553320, receiver=js::MaybeRooted<JS::Value, js::AllowGC>::HandleType at 0x000000013b553318, id=js::MaybeRooted<jsid, js::AllowGC>::HandleType at 0x000000013b553310, nameLookup=NotNameLookup, vp=js::MaybeRooted<JS::Value, js::AllowGC>::MutableHandleType at 0x000000013b553308)1>(JSContext*, js::MaybeRooted<js::NativeObject*, (js::AllowGC)1>::HandleType, js::MaybeRooted<JS::Value, (js::AllowGC)1>::HandleType, js::MaybeRooted<jsid, (js::AllowGC)1>::HandleType, IsNameLookup, js::MaybeRooted<JS::Value, (js::AllowGC)1>::MutableHandleType) + 458 at NativeObject.cpp:2002
    frame #19: 0x00000001033a15dd servo`js::NativeGetProperty(cx=0x000000010b0d5240, obj=js::HandleNativeObject at 0x000000013b553388, receiver=JS::HandleValue at 0x000000013b553380, id=JS::HandleId at 0x000000013b553378, vp=JS::MutableHandleValue at 0x000000013b553370) + 93 at NativeObject.cpp:2036
    frame #20: 0x00000001030ff596 servo`js::GetProperty(cx=0x000000010b0d5240, obj=JS::HandleObject at 0x000000013b553410, receiver=JS::HandleValue at 0x000000013b553408, id=JS::HandleId at 0x000000013b553400, vp=JS::MutableHandleValue at 0x000000013b5533f8) + 214 at NativeObject.h:1477
    frame #21: 0x000000010314a328 servo`js::GetProperty(cx=0x000000010b0d5240, obj=JS::HandleObject at 0x000000013b5534a8, receiver=JS::HandleValue at 0x000000013b5534a0, name=0x000000013d181880, vp=JS::MutableHandleValue at 0x000000013b553498) + 136 at jsobj.h:822
    frame #22: 0x00000001033695d5 servo`js::GetProperty(cx=0x000000010b0d5240, v=JS::HandleValue at 0x000000013b5535b0, name=js::HandlePropertyName at 0x000000013b5535a8, vp=JS::MutableHandleValue at 0x000000013b5535a0) + 933 at Interpreter.cpp:4087
    frame #23: 0x0000000103372e67 servo`GetPropertyOperation(cx=0x000000010b0d5240, fp=0x000000010baa2a78, script=JS::HandleScript at 0x000000013b5536a0, pc=0x000000011005f539, lval=JS::MutableHandleValue at 0x000000013b553698, vp=JS::MutableHandleValue at 0x000000013b553690) + 535 at Interpreter.cpp:217
    frame #24: 0x0000000103354ce0 servo`Interpret(cx=0x000000010b0d5240, state=0x000000013b5567f8) + 47872 at Interpreter.cpp:2528
    frame #25: 0x00000001033490df servo`js::RunScript(cx=0x000000010b0d5240, state=0x000000013b5567f8) + 799 at Interpreter.cpp:426
    frame #26: 0x000000010333b9eb servo`js::Invoke(cx=0x000000010b0d5240, args=0x000000013b556e50, construct=NO_CONSTRUCT) + 1307 at Interpreter.cpp:494
    frame #27: 0x00000001031593a3 servo`js::fun_apply(cx=0x000000010b0d5240, argc=2, vp=0x000000010baa2a50) + 1667 at jsfun.cpp:1232
    frame #28: 0x000000010337c24b servo`js::CallJSNative(cx=0x000000010b0d5240, native=0x0000000103158d20, args=0x000000013b5590b0)(JSContext*, unsigned int, JS::Value*), JS::CallArgs const&) + 187 at jscntxtinlines.h:235
    frame #29: 0x000000010333b913 servo`js::Invoke(cx=0x000000010b0d5240, args=0x000000013b5590b0, construct=NO_CONSTRUCT) + 1091 at Interpreter.cpp:476
    frame #30: 0x0000000103357dc3 servo`Interpret(cx=0x000000010b0d5240, state=0x000000013b55a198) + 60387 at Interpreter.cpp:2809
    frame #31: 0x00000001033490df servo`js::RunScript(cx=0x000000010b0d5240, state=0x000000013b55a198) + 799 at Interpreter.cpp:426
    frame #32: 0x000000010333b9eb servo`js::Invoke(cx=0x000000010b0d5240, args=0x000000013b55a2d0, construct=NO_CONSTRUCT) + 1307 at Interpreter.cpp:494
    frame #33: 0x000000010336682d servo`js::Invoke(cx=0x000000010b0d5240, thisv=0x000000013b55a3d0, fval=0x000000013b55a6f0, argc=1, argv=0x000000012683c178, rval=JS::MutableHandleValue at 0x000000013b55a2c0) + 621 at Interpreter.cpp:528
    frame #34: 0x00000001030df88b servo`JS_CallFunctionValue(cx=0x000000010b0d5240, obj=JS::HandleObject at 0x000000013b55a418, fval=JS::HandleValue at 0x000000013b55a410, args=0x000000013b55a5c0, rval=JS::MutableHandleValue at 0x000000013b55a408) + 347 at jsapi.cpp:2852
    frame #35: 0x000000010185eaa6 servo`script::dom::bindings::codegen::Bindings::EventListenerBinding::{{impl}}::HandleEvent(self=0x0000000118a85370, cx=0x000000010b0d5240, aThisObj=Handle<*mut js::jsapi::JSObject> at 0x000000013b55a800, event=0x0000000118a12f00) + 2342 at EventListenerBinding.rs:155
    frame #36: 0x0000000101e916cf servo`script::dom::bindings::codegen::Bindings::EventListenerBinding::{{impl}}::HandleEvent_<script::dom::eventtarget::EventTarget>(self=0x0000000118a85370, thisObj=0x0000000126862000, event=0x0000000118a12f00, aExceptionHandling=Report) + 655 at EventListenerBinding.rs:125
    frame #37: 0x0000000101e903af servo`script::dom::eventtarget::{{impl}}::call_or_handle_event<script::dom::eventtarget::EventTarget>(self=0x0000000126810120, object=0x0000000126862000, event=0x0000000118a12f00, exception_handle=Report) + 399 at eventtarget.rs:151
    frame #38: 0x0000000101e9015f servo`script::dom::eventdispatcher::handle_event(window=Option<&script::dom::window::Window> at 0x000000013b55b360, listener=0x0000000126810120, current_target=0x0000000126862000, event=0x0000000118a12f00) + 255 at eventdispatcher.rs:47
    frame #39: 0x0000000101e968e8 servo`script::dom::eventdispatcher::inner_invoke(window=Option<&script::dom::window::Window> at 0x000000013b55b428, object=0x0000000126862000, event=0x0000000118a12f00, listeners=&[script::dom::eventtarget::CompiledEventListener] at 0x000000013b55b408) + 280 at eventdispatcher.rs:215
    frame #40: 0x0000000101e932a6 servo`script::dom::eventdispatcher::invoke(window=Option<&script::dom::window::Window> at 0x000000013b55b560, object=0x0000000126862000, event=0x0000000118a12f00, specific_listener_phase=Option<script::dom::eventtarget::ListenerPhase> at 0x000000013b55b548) + 598 at eventdispatcher.rs:190
    frame #41: 0x0000000101e92970 servo`script::dom::eventdispatcher::dispatch_to_listeners(event=0x0000000118a12f00, target=0x0000000126862000, event_path=&[&script::dom::eventtarget::EventTarget] at 0x000000013b55b838) + 1520 at eventdispatcher.rs:84
    frame #42: 0x0000000101e9412d servo`script::dom::eventdispatcher::dispatch_event(target=0x0000000126862000, target_override=Option<&script::dom::eventtarget::EventTarget> at 0x000000013b55be10, event=0x0000000118a12f00) + 3341 at eventdispatcher.rs:148
    frame #43: 0x000000010162b504 servo`script::dom::eventtarget::{{impl}}::dispatch_event(self=0x0000000126862000, event=0x0000000118a12f00) + 36 at eventtarget.rs:308
    frame #44: 0x0000000101da55fd servo`script::dom::event::{{impl}}::fire(self=0x0000000118a12f00, target=0x0000000126862000) + 45 at event.rs:297
    frame #45: 0x0000000101ea4907 servo`script::dom::eventtarget::{{impl}}::fire_event(self=0x0000000126862000, name=&str at 0x000000013b55bf60, bubbles=Bubbles, cancelable=NotCancelable) + 263 at eventtarget.rs:502
    frame #46: 0x00000001022395d1 servo`script::task_source::dom_manipulation::{{impl}}::handle_task(self=DOMManipulationTask at 0x000000013b55c1a8, script_thread=0x000000013b55ff80) + 1249 at dom_manipulation.rs:54
    frame #47: 0x00000001021a8080 servo`script::script_thread::{{impl}}::handle_msg_from_script(self=0x000000013b55ff80, msg=MainThreadScriptMsg at 0x000000013b55c4e8) + 688 at script_thread.rs:941
    frame #48: 0x0000000102230082 servo`script::script_thread::{{impl}}::handle_msgs::{{closure}} + 434 at script_thread.rs:771
    frame #49: 0x000000010222fbe1 servo`script::script_thread::{{impl}}::profile_event<closure,core::option::Option<bool>>(self=0x000000013b55ff80, category=ScriptEvent, f=closure at 0x000000013b55ce98) + 1009 at script_thread.rs:863
    frame #50: 0x0000000102201186 servo`script::script_thread::{{impl}}::handle_msgs(self=0x000000013b55ff80) + 7734 at script_thread.rs:763
    frame #51: 0x00000001021a5d5d servo`script::script_thread::{{impl}}::start(self=0x000000013b55ff80) + 29 at script_thread.rs:615
    frame #52: 0x00000001021a5d15 servo`script::script_thread::{{impl}}::create::{{closure}}::{{closure}} + 37 at script_thread.rs:462
    frame #53: 0x00000001021a5a62 servo`profile_traits::mem::{{impl}}::run_with_memory_reporting<closure,fn(self=0x000000013b5601d0, f=closure at 0x000000013b55f9e0, reporter_name=String at 0x000000013b55f9c8, channel_for_reporter=Sender<script::script_thread::MainThreadScriptMsg> at 0x000000013b55f9b0, msg=0x0000000000000015) -> script::script_runtime::CommonScriptMsg,script::script_runtime::CommonScriptMsg,std::sync::mpsc::Sender<script::script_thread::MainThreadScriptMsg>> + 674 at mem.rs:59
    frame #54: 0x0000000102192cba servo`script::script_thread::{{impl}}::create::{{closure}} + 1882 at script_thread.rs:461
    frame #55: 0x0000000102191251 servo`util::thread::spawn_named_with_send_on_panic::_$u7b$$u7b$closure$u7d$$u7d$::h51879a609c3205a6 + 561 at boxed.rs:234
    frame #56: 0x0000000102190fc8 servo`std::panic::{{impl}}::call_once<(self=AssertUnwindSafe<closure> at 0x000000013b560780, _args=<unavailable>),closure> + 136 at panic.rs:284
    frame #57: 0x0000000102190f1f servo`std::panicking::try::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h89eb0322b68ea443 + 79
    frame #58: 0x00000001021917f5 servo`std::panicking::try::call<closure>(f=0x000000013b560ee0) + 21 at panicking.rs:272
    frame #59: 0x0000000103c64d3c servo`__rust_try + 12
    frame #60: 0x0000000103c64cd6 servo`__rust_maybe_catch_panic + 38
    frame #61: 0x0000000102190e6e servo`std::panicking::try::_$u7b$$u7b$closure$u7d$$u7d$::hb39ecbfd724b9507 + 174
    frame #62: 0x0000000102190db0 servo`std::thread::local::{{impl}}::with<core::cell::Cell<usize>,closure,core::result::Result<(self=0x000000010501b020, f=closure at 0x000000013b560f78), Box<Any>>> + 208 at local.rs:211
    frame #63: 0x0000000102190be0 servo`std::panicking::try<(f=AssertUnwindSafe<closure> at 0x000000013b561250),std::panic::AssertUnwindSafe<closure>> + 192 at panicking.rs:235
    frame #64: 0x0000000102190abf servo`std::panic::catch_unwind<std::panic::AssertUnwindSafe<closure>,(f=AssertUnwindSafe<closure> at 0x000000013b5616e0)> + 143 at panic.rs:387
    frame #65: 0x000000010219092f servo`std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::h9c63aaed83bdcff4 + 383
    frame #66: 0x00000001021919d8 servo`alloc::boxed::{{impl}}::call_box<(self=0x000000011b63a780, args=<unavailable>),closure> + 104 at boxed.rs:543
    frame #67: 0x0000000103c609f9 servo`std::sys::thread::Thread::new::thread_start::h6f266e069bf4ec2b + 57
    frame #68: 0x00007fff84e35899 libsystem_pthread.dylib`_pthread_body + 138
    frame #69: 0x00007fff84e3572a libsystem_pthread.dylib`_pthread_start + 137
    frame #70: 0x00007fff84e39fc9 libsystem_pthread.dylib`thread_start + 13
```

**Comment by jdm** [comment]
I did it a second time after waiting for a lot longer and saw the same backtrace. We appear to be stuck waiting on a layout mutex.

**Comment by jdm** [comment]
Meanwhile, the layout thread for that page doesn't appear to be doing anything interesting:
```
(lldb) bt
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
* thread #102: tid = 0x9f589, 0x00007fff8e230716 libsystem_kernel.dylib`__psynch_cvwait + 10
  * frame #0: 0x00007fff8e230716 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff84e37c3b libsystem_pthread.dylib`_pthread_cond_wait + 727
    frame #2: 0x0000000103c47fa9 servo`std::thread::park::h73cb11855e7141e1 + 297
    frame #3: 0x0000000103c57f10 servo`std::sync::mpsc::select::Select::wait::h0b444db680dca85f + 288
    frame #4: 0x0000000100fdbee3 servo`layout::layout_thread::{{impl}}::handle_request(self=0x000000013b4f6a38, possibly_locked_rw_data=0x000000013b4f69c8) + 611 at <std macros>:6
    frame #5: 0x0000000100fa1a42 servo`layout::layout_thread::{{impl}}::start(self=LayoutThread at 0x000000013b4f6a38) + 370 at layout_thread.rs:494
    frame #6: 0x0000000100fa1888 servo`layout::layout_thread::{{impl}}::create::{{closure}}::{{closure}} + 104 at layout_thread.rs:287
    frame #7: 0x0000000100f9b7d0 servo`profile_traits::mem::{{impl}}::run_with_memory_reporting<closure,fn(self=0x000000013b4f7dd0, f=closure at 0x000000013b4f7278, reporter_name=String at 0x000000013b4f7260, channel_for_reporter=Sender<script::layout_interface::Msg> at 0x000000013b4f7248, msg=0x000000013b4f7280) -> script::layout_interface::Msg,script::layout_interface::Msg,std::sync::mpsc::Sender<script::layout_interface::Msg>> + 912 at mem.rs:59
    frame #8: 0x0000000100f7e9f3 servo`layout::layout_thread::{{impl}}::create::{{closure}} + 2323 at layout_thread.rs:286
    frame #9: 0x0000000100f7a7a1 servo`util::thread::spawn_named_with_send_on_panic::_$u7b$$u7b$closure$u7d$$u7d$::h34b295e330d3a42d + 561 at boxed.rs:234
    frame #10: 0x0000000100f7a518 servo`std::panic::{{impl}}::call_once<(self=AssertUnwindSafe<closure> at 0x000000013b4f7ff0, _args=<unavailable>),closure> + 136 at panic.rs:284
    frame #11: 0x0000000100f7a46f servo`std::panicking::try::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h8d2fe62ffa2338ab + 79
    frame #12: 0x0000000100f7ad45 servo`std::panicking::try::call<closure>(f=0x000000013b4f8480) + 21 at panicking.rs:272
    frame #13: 0x0000000103c64d3c servo`__rust_try + 12
    frame #14: 0x0000000103c64cd6 servo`__rust_maybe_catch_panic + 38
    frame #15: 0x0000000100f7a15e servo`std::panicking::try::_$u7b$$u7b$closure$u7d$$u7d$::h0508309180bde640 + 174
    frame #16: 0x0000000100f7a030 servo`std::thread::local::{{impl}}::with<core::cell::Cell<usize>,closure,core::result::Result<(self=0x000000010501b020, f=closure at 0x000000013b4f8518), Box<Any>>> + 208 at local.rs:211
    frame #17: 0x0000000100f79e60 servo`std::panicking::try<(f=AssertUnwindSafe<closure> at 0x000000013b4f8700),std::panic::AssertUnwindSafe<closure>> + 192 at panicking.rs:235
    frame #18: 0x0000000100f79d3f servo`std::panic::catch_unwind<std::panic::AssertUnwindSafe<closure>,(f=AssertUnwindSafe<closure> at 0x000000013b4f89b0)> + 143 at panic.rs:387
    frame #19: 0x0000000100f79baf servo`std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::h226a550ab4accc2b + 383
    frame #20: 0x0000000100f7af88 servo`alloc::boxed::{{impl}}::call_box<(self=0x000000011b035640, args=<unavailable>),closure> + 104 at boxed.rs:543
    frame #21: 0x0000000103c609f9 servo`std::sys::thread::Thread::new::thread_start::h6f266e069bf4ec2b + 57
    frame #22: 0x00007fff84e35899 libsystem_pthread.dylib`_pthread_body + 138
    frame #23: 0x00007fff84e3572a libsystem_pthread.dylib`_pthread_start + 137
    frame #24: 0x00007fff84e39fc9 libsystem_pthread.dylib`thread_start + 13
```

**Comment by jdm** [comment]
Interestingly, despite the script thread having returned from the `reflow` call (which is used to compute the query results), there does not appear to be any layout information stored in the layout data that's being locked. One theory being floated right now is that this is occurring before the first layout of the page, which causes the layout data to [blocked](https://github.com/nox/servo/blob/7b467ee52d59d1d2f078948acc9e4d188eac7338/components/layout/layout_thread.rs#L345-L354).

**Comment by jdm** [comment]
The layout  thread does have `first_reflow` set to `true`, which means that it has not executed `perform_post_main_layout_passes`.

**Comment by jdm** [comment]
The reflow doesn't happen because `Window::window_size` is None, so `force_reflow` bails early on and never sends a message to the layout thread.

**Comment by emilio** [comment]
Also, while trying to test this, I found out we deadlock in a very simple example:

```html
<!doctype html>                                                                             
<script>                                                                                    
var frame;                                                                                                                                                                              
setTimeout(function f() {                                                                   
  if (frame)                                                                                
    frame.parentNode.removeChild(frame);                                                    
                                                                                            
  frame = document.createElement('iframe');                                                 
  frame.src = "frame.html";                                                                 
  frame.style.display = "none";                                                             
  document.body.appendChild(frame);                                                         
  setTimeout(f, 150);                                                                       
}, 150)                                                                                     
</script>
```

(We never manage to remove the first frame, which is worrying)

**Comment by emilio** [comment]
@jdm: So it seems to reproduce the original error we need a chunked response or something like that, right?

**Comment by jdm** [comment]
Yeah, I assume there's a timing issue where small responses end up getting parsed too quickly or something? I've been having trouble figuring out how the window size messages would end up being delayed, though.

**Comment by emilio** [comment]
So the only moment when there can be no window_size is while there's an
incomplete load, or if for some reason when the event arrives the
browsing context isn't correctly set up and we can't find the correct
pipeline to deliver the event to...

Could it be the case that while the page is loading, a parent script is
trying to query the iframe contents via `contentDocument` or similar,
but the viewport is still not correctly setup in the child?

**Comment by nox** [comment]
Does it even make sense to do queries for offsets and whatnot when there is no `window_size`? How wrong would it be to return `(None, Rect::zero())` in that case?

**Comment by emilio** [comment]
So I was able to reproduce it consistently:

Take the following `test.html` file:

```

**Comment by paulrouget** [comment]
Thank you all for figuring out what was going on! It now works perfectly.
